Fritz.Twitch/Proxy.cs

```diff
@@ -71,27 +78,63 @@ private void ConfigureClient(HttpClient client)
 
 		private async Task<HttpResponseMessage> GetFromEndpoint(string url)
 		{
-
-			// Check rate-limit
-			_RateLimitLock.EnterReadLock();
-			if (_RateLimitRemaining <= 0)
+			var isQueued = false;
+			do
 			{
-				_RateLimitLock.ExitReadLock();
-				await Task.Delay(_RateLimitReset.Subtract(DateTime.UtcNow));
-				return await GetFromEndpoint(url);
+				// Check rate-limit
+				lock (_QueueLock)
+				{
+					if (_RateLimitRemaining - _WaitingRequests > 0)
+					{
+						_WaitingRequests++;
+						if (isQueued)
+						{
+							_QueuedRequests--;
+							isQueued = false;
+						}
+					}
+					else
+					{
+						if (!isQueued)
+						{
+							if (_QueuedRequests + 1 > MAX_QUEUED)
+							{
+								throw new TimeoutException("Too many requests waiting");
+							}
+							_QueuedRequests++;
+							isQueued = true;
+						}
+					}
+				}
+
+				if (isQueued)
+				{
+					await Task.Delay(new DateTime(Volatile.Read(ref _RateLimitResetTicks)).Subtract(DateTime.UtcNow));
+				}
 			}
-			_RateLimitLock.ExitReadLock();
+			while (isQueued);
 
-			var result = await Client.GetAsync(url);
+			HttpResponseMessage result;
+			await _RateLimitLock.WaitAsync();
+			try
+			{
+				result = await Client.GetAsync(url);
 
-			var remaining = short.Parse(result.Headers.GetValues("RateLimit-Remaining").First());
-			var reset = long.Parse(result.Headers.GetValues("RateLimit-Reset").First());
+				var remaining = short.Parse(result.Headers.GetValues("RateLimit-Remaining").First());
+				var reset = long.Parse(result.Headers.GetValues("RateLimit-Reset").First());
 
-			_RateLimitLock.EnterWriteLock();
-			_RateLimitRemaining = remaining;
-			_RateLimitReset = reset.ToDateTime();
-			Logger.LogTrace($"{DateTime.UtcNow}: Twitch Rate - {remaining} until {_RateLimitReset}");
-			_RateLimitLock.ExitWriteLock();
+				lock (_QueueLock)
+				{
+					_RateLimitRemaining = remaining;
+					_WaitingRequests--;
+					Volatile.Write(ref _RateLimitResetTicks, reset.ToDateTime().Ticks);
```

Comment:
Why Volatile.Write and not Interlocked...?

Reply:
Which Interlocked?

Reply:
Interlocked.Exchange

Reply:
Don't care about the previous value (just want to overwrite it); so don't need to exchange